Add dependency injection to KeyManager

mradcliffe opened the pull request:
This branch changes the KeyManager service to use dependency injection, and adds PHP Unit tests.

There are two failing tests due to code errors inside KeyManager:

1. getKeys() specifies a string argument but loadMultiple takes an array argument.
2. getDefaultKeyContents() uses a method getKeyContents but this does not exist.

I added phpunit.xml inside the module directory which is configured to use Drupal bootstrap like how I do it with [Xero](https://github.com/mradcliffe/xero). This allows to run unit tests like so:

**Updated**

Now uses drupal-ti to run both simpletest and phpunit tests.

https://travis-ci.org/mradcliffe/key/builds/73238879

nerdstein commented:
Thanks. I'll take a look tomorrow.

nerdstein commented on key.services.yml:
```diff
@@ -2,7 +2,8 @@ services:
 
   key_manager:
     class: Drupal\key\KeyManager
+    arguments: ['@entity.manager', '@config.factory']
```
Is there any changes we should make to the readme on how to invoke the service?

mradcliffe replied:
That is the beauty of the service container. The service container handles this whether invoked by \Drupal::service or injected into another class. It also abstracts \Drupal away from the class so that theoretically someone could use it outside of the Drupal stack if they had some services for entity.manager and config.factory that behaved similarly.

nerdstein replied:
Cool, thanks for the explanation. Do we need to update the documentation in this repo's readme.md file?

nerdstein commented on phpunit.xml:
```diff
@@ -0,0 +1,34 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Is there a convenient way to integrate this with Github and show code coverage there?

mradcliffe replied:
[Coveralls](https://coveralls.io) is one option, but I have had difficulty implementing a nice solution for Drupal modules via TravisCI without running code coverage for every PHP version. It also puts a composer dependency into the module on php-coveralls.

What I am currently doing is running phpunit with --coverage-text, and printing out that report in TravisCI, but that will not directly report back to GitHub what the new coverage is. TravisCI will allow to run the tests automatically for PHP Unit tests.

A sample .travis.yml file:
```yaml
language: php

php:
  - 5.5
  - 5.6

sudo: false

install:
  # Download Drupal.
  - TESTDIR=$(pwd)
  - cd ..
  - git clone --branch 8.0.x --depth 1 http://git.drupal.org/project/drupal.git drupal
  - cd drupal

before_script:
  # rsync the module directory into modules/
  - rsync -rtlDPvc --exclude .git/ $TESTDIR modules/
  - cd modules/key
script:
  - ../../core/vendor/bin/phpunit --coverage-text
```

nerdstein replied:
Can you commit a travis.yml file and a composer.json with PHPUnit? I think we can load the composer libs within Travis and get the same functionality

nerdstein commented on phpunit.xml:
```diff
@@ -0,0 +1,34 @@
+  </php>
+  <testsuites>
+    <testsuite name="Key Module Unit Test Suite">
+      <directory>tests</directory>
```
Just to be clear, this *only* targets Key tests and not all tests?

mradcliffe replied:
Yes, phpunit will use the configuration file it finds in the current working directory. I ran into some cases where phpunit would try to run all of tests it finds in the current working directory.

nerdstein replied:
Great

nerdstein commented on tests/src/KeyManagerTest.php:
```diff
@@ -0,0 +1,136 @@
+use Drupal\key\Entity\Key;
+use Drupal\key\KeyManager;
+
+class KeyManagerTest extends KeyTestBase {
```
Do you know we also have src/Tests for SimpleTests?

This seems like a parallel but related set of tests. Can we somehow merge them and get the same results?

mradcliffe replied:
Yes, there is some overlap between functional and unit tests, but there are some differences.

1. Unit tests test the individual class in isolation without having to worry about state. A unit test should assert valid output for a method for each code path.
   1. This also means that unit tests are faster because it does not depend on system state. This is hard to do in practice in Drupal 8 because of the amount of mocking that has to be done, but those tests ran in less than in a second. :+1: 
2. Functional tests are good for testing the output from an actual browser, but the downside is that it takes longer to run because we need to spin up a database test.

I think as Drupal 8.1.x evolves there will be less functional tests when unit tests will suffice. This would reduce the testing infrastructure burden and speed up feedback on patches.

The phpunit test directory standard is to be in `tests` and the SimpleTest directory in `src/tests`. This is seen in Core as well.

nerdstein replied:
Perfect. OK

nerdstein commented:
Should we consider adding a composer.json file for any needed code libraries? e.g. PHPUnit

nerdstein commented:
I've reviewed the code and made some feedback. Overall, this PR looks great. One major missing piece is the readme. I'd like to make sure that your efforts are captured there so others know how to leverage the PHPUnit work you have done.

mradcliffe commented:
Regarding composer.json, at the moment key does not need any dependency management as it uses what is in core. This does mean that having core checked out is required for those tests (mainly for ContainerBuilder actually). I don't think that Drupal will be at the point where modules are self-contained like Bundles or Laravel libraries. 

I will take a look at documentation... Probably tomorrow.

mradcliffe commented:
Just as an example I added my fork to my travis account: https://travis-ci.org/mradcliffe/key/jobs/73181262

And added some information regarding dependency injection to the readme.

mradcliffe commented:
Actually, if we follow https://github.com/commerceguys/commerce/blob/8.x-2.x/.travis.yml, then it's possible to run the simpletest as well. It will just take a lot longer. :(

The phpunit tests ran in like seconds while if you look at Commerce's simpletest integration, it takes about 15 minutes. On d.o, it would probably take less to run simpletest depending on Jenkins queue time. That will probably take a .drupalci.yml file to specifically run phpunit tests for drupalci iirc.

nerdstein commented:
I just read over all comments, let me summarize next steps to consider:

1. Create composer.json *for this project* that has PHPUnit (this runs independently to core)
2. Enhance travis to invoke composer libraries, run simpletests, and run phpunit (note: I am OK with the time lag because of the added benefit)

I'd be thrilled if we can make that happen and then get this merged in

nerdstein commented:
@mradcliffe - does this section of the readme.md need updated?

https://github.com/nerdstein/key#use-of-services

mradcliffe commented:
1. **Re: Readme.md**
   * I added some links to explanations about how to leverage dependency injection in Drupal 8 to the readme.md file in a prior commit, and I think that is sufficient.
2. **Re: Composer.json**
   * We don't need to include phpunit in composer.json because Drupal already includes it as part of core. Furthermore apparently the drupal-ti test runner also includes phpunit.
3. **Re: Travis**
   * I just merged a commit to change to using [drupal-ti](https://github.com/LionsAd/drupal_ti/), which identified #13 and I fixed that. Here's the build for that commit: https://travis-ci.org/mradcliffe/key/builds/73238879. It's a failed build because of the bugs I wrote tests for that are still open.
   * The next step after this PR would be to create an open source [Travis CI](http://www.travis-ci.org) account and associate this repository with it.

nerdstein commented:
Awesome. I'll review this tonight and try to get this Travis account set up

nerdstein commented:
Merging
